libnvme/libnvme/libnvme_ctrl_spec.h

```diff
@@ -0,0 +1,1498 @@
+/*
+ * Copyright (C) 2017-2019 Red Hat, Inc.
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Gris Ge <fge@redhat.com>
+ */
+
+#ifndef _LIBNVME_CTRL_SPEC_H_
+#define _LIBNVME_CTRL_SPEC_H_
+
+#include "libnvme_common.h"
+
+#include <stdint.h>
+
+#define NVME_SPEC_CTRL_RAW_IDENTIFY_DATA_LEN		4096
+#define NVME_SPEC_CTRL_VENDOR_SPECFIC_DATA_LEN		1024
+#define NVME_SPEC_CTRL_SN_LEN				20
+#define NVME_SPEC_CTRL_MN_LEN				40
+#define NVME_SPEC_CTRL_FR_LEN				8
+#define NVME_SPEC_CTRL_SUBNQN_LEN			256
+#define NVME_SPEC_MXPS_001W				0
+#define NVME_SPEC_MXPS_00001W				1
+#define NVME_SPEC_IPS_NO_REPORT				0b00
+#define NVME_SPEC_IPS_00001W				0b01
+#define NVME_SPEC_IPS_001W				0b10
+#define NVME_SPEC_APS_NO_REPORT				0b00
+#define NVME_SPEC_APS_00001W				0b01
+#define NVME_SPEC_APS_001W				0b10
+
+struct _DLL_PUBLIC nvme_ctrl;
+struct _DLL_PUBLIC nvme_psd;
+
+/**
+ * nvme_ctrl_vid_get() - Retrieve the VID property of specified NVMe controller.
+ *
+ * Retrieve the VID(Vendor ID) property of specified NVMe controller.
+ *
+ * @cnt:
+ *	Pointer of 'struct nvme_ctrl'.
+ *	If this pointer is NULL, your program will be terminated by assert().
```

Comment:
In general I think it's frowned upon to assert in a library.  Remember however that:
```
If the macro NDEBUG is defined at the moment <assert.h> was last included, the macro assert()
generates no code, and hence does nothing at all.
```